File: crates/viben-core/src/models/discovery.rs
//! Model discovery for various providers

use crate::error::{Error, Result};
use crate::providers::{get_default_base_url, Provider, ProviderType};
use serde::Deserialize;

use super::types::DiscoveredModel;
use super::known::get_known_models_for_provider;

/// Discover models available from a provider
pub async fn discover_models(provider: &Provider) -> Result<Vec<DiscoveredModel>> {
    match provider.provider_type {
        ProviderType::OpenAI => discover_openai_models(provider).await,
        ProviderType::Anthropic => discover_anthropic_models(provider).await,
        ProviderType::Ollama => discover_ollama_models(provider).await,
        ProviderType::OpenRouter => discover_openrouter_models(provider).await,
        ProviderType::Google => discover_google_models(provider).await,
        ProviderType::Azure | ProviderType::Custom => {
            // Azure and Custom providers don't have a standard discovery API
            Ok(vec![])
        }
    }
}

/// Discover OpenAI models via API
/// GET {base_url}/models with Bearer token auth
async fn discover_openai_models(provider: &Provider) -> Result<Vec<DiscoveredModel>> {
    let base_url = provider
        .base_url
        .as_deref()
        .or(get_default_base_url(ProviderType::OpenAI))
        .ok_or_else(|| Error::ModelDiscovery("No base URL for OpenAI provider".to_string()))?;

    let api_key = provider
        .api_key
        .as_ref()
        .ok_or_else(|| Error::ModelDiscovery("No API key for OpenAI provider".to_string()))?;

    let url = format!("{}/models", base_url.trim_end_matches('/'));

    let client = reqwest::Client::new();
    let response = client
        .get(&url)
        .header("Authorization", format!("Bearer {}", api_key))
        .send()
        .await?;

    if !response.status().is_success() {
        let status = response.status();
        let text = response.text().await.unwrap_or_default();
        return Err(Error::ModelDiscovery(format!(
            "OpenAI API error ({}): {}",
            status, text
        )));
    }

    #[derive(Deserialize)]
    struct OpenAIModel {
        id: String,
        #[serde(default)]
        owned_by: Option<String>,
        #[serde(default)]
        created: Option<i64>,
    }

    #[derive(Deserialize)]
    struct OpenAIModelsResponse {
        data: Vec<OpenAIModel>,
    }

    let models_response: OpenAIModelsResponse = response.json().await?;

    // Filter to only include chat models (gpt-*, o1-*, o3-*, etc.)
    let models: Vec<DiscoveredModel> = models_response
        .data
        .into_iter()
        .filter(|m| {
            let id = m.id.to_lowercase();
            id.starts_with("gpt-")
                || id.starts_with("o1")
                || id.starts_with("o3")
                || id.starts_with("chatgpt")
        })
        .map(|m| DiscoveredModel {
            id: m.id.clone(),
            name: m.id.clone(),
            description: None,
            context_window: None,
            max_output_tokens: None,
            owned_by: m.owned_by,
            created: m.created,
        })
        .collect();

    Ok(models)
}

/// Discover Anthropic models
/// Anthropic doesn't have a models listing API, use known models
async fn discover_anthropic_models(_provider: &Provider) -> Result<Vec<DiscoveredModel>> {
    let known = get_known_models_for_provider(ProviderType::Anthropic);
    let models: Vec<DiscoveredModel> = known
        .into_iter()
        .map(|m| DiscoveredModel {
            id: m.id.to_string(),
            name: m.name.to_string(),
            description: m.description.map(|s| s.to_string()),
            context_window: m.context_window,
            max_output_tokens: m.max_output_tokens,
            owned_by: Some("anthropic".to_string()),
            created: None,
        })
        .collect();
    Ok(models)
}

/// Discover Ollama models via API
/// GET {base_url}/api/tags (no auth)
async fn discover_ollama_models(provider: &Provider) -> Result<Vec<DiscoveredModel>> {
    let base_url = provider
        .base_url
        .as_deref()
        .or(get_default_base_url(ProviderType::Ollama))
        .ok_or_else(|| Error::ModelDiscovery("No base URL for Ollama provider".to_string()))?;

    let url = format!("{}/api/tags", base_url.trim_end_matches('/'));

    let client = reqwest::Client::new();
    let response = client.get(&url).send().await;

    // Ollama might not be running, return empty list instead of error
    let response = match response {
        Ok(r) => r,
        Err(_) => return Ok(vec![]),
    };

    if !response.status().is_success() {
        return Ok(vec![]);
    }

    #[derive(Deserialize)]
    struct OllamaModel {
        name: String,
    }

    #[derive(Deserialize)]
    struct OllamaModelsResponse {
        models: Vec<OllamaModel>,
    }

    let models_response: OllamaModelsResponse = match response.json().await {
        Ok(r) => r,
        Err(_) => return Ok(vec![]),
    };

    let models: Vec<DiscoveredModel> = models_response
        .models
        .into_iter()
        .map(|m| DiscoveredModel {
            id: m.name.clone(),
            name: m.name,
            description: None,
            context_window: None,
            max_output_tokens: None,
            owned_by: Some("local".to_string()),
            created: None,
        })
        .collect();

    Ok(models)
}

/// Discover OpenRouter models via API
/// GET https://openrouter.ai/api/v1/models with Bearer token
async fn discover_openrouter_models(provider: &Provider) -> Result<Vec<DiscoveredModel>> {
    let base_url = provider
        .base_url
        .as_deref()
        .or(get_default_base_url(ProviderType::OpenRouter))
        .ok_or_else(|| Error::ModelDiscovery("No base URL for OpenRouter provider".to_string()))?;

    let url = format!("{}/models", base_url.trim_end_matches('/'));

    let client = reqwest::Client::new();
    let mut request = client.get(&url);

    // OpenRouter allows unauthenticated requests to /models
    if let Some(api_key) = &provider.api_key {
        request = request.header("Authorization", format!("Bearer {}", api_key));
    }

    let response = request.send().await?;

    if !response.status().is_success() {
        let status = response.status();
        let text = response.text().await.unwrap_or_default();
        return Err(Error::ModelDiscovery(format!(
            "OpenRouter API error ({}): {}",
            status, text
        )));
    }

    #[derive(Deserialize)]
    struct OpenRouterModel {
        id: String,
        name: String,
        #[serde(default)]
        description: Option<String>,
        #[serde(default)]
        context_length: Option<u32>,
    }

    #[derive(Deserialize)]
    struct OpenRouterModelsResponse {
        data: Vec<OpenRouterModel>,
    }

    let models_response: OpenRouterModelsResponse = response.json().await?;

    let models: Vec<DiscoveredModel> = models_response
        .data
        .into_iter()
        .map(|m| {
            // Extract owner from model ID (e.g., "openai/gpt-4" -> "openai")
            let owned_by = m.id.split('/').next().map(|s| s.to_string());
            DiscoveredModel {
                id: m.id,
                name: m.name,
                description: m.description,
                context_window: m.context_length,
                max_output_tokens: None,
                owned_by,
                created: None,
            }
        })
        .collect();

    Ok(models)
}

/// Discover Google AI (Gemini) models via API
/// GET {base_url}/models?key={api_key}
async fn discover_google_models(provider: &Provider) -> Result<Vec<DiscoveredModel>> {
    let base_url = provider
        .base_url
        .as_deref()
        .or(get_default_base_url(ProviderType::Google))
        .ok_or_else(|| Error::ModelDiscovery("No base URL for Google provider".to_string()))?;

    let api_key = provider
        .api_key
        .as_ref()
        .ok_or_else(|| Error::ModelDiscovery("No API key for Google provider".to_string()))?;

    let url = format!("{}/models?key={}", base_url.trim_end_matches('/'), api_key);

    let client = reqwest::Client::new();
    let response = client.get(&url).send().await?;

    if !response.status().is_success() {
        let status = response.status();
        let text = response.text().await.unwrap_or_default();
        return Err(Error::ModelDiscovery(format!(
            "Google AI API error ({}): {}",
            status, text
        )));
    }

    #[derive(Deserialize)]
    #[serde(rename_all = "camelCase")]
    struct GoogleModel {
        name: String,
        #[serde(default)]
        display_name: Option<String>,
        #[serde(default)]
        description: Option<String>,
        #[serde(default)]
        input_token_limit: Option<u32>,
        #[serde(default)]
        output_token_limit: Option<u32>,
    }

    #[derive(Deserialize)]
    struct GoogleModelsResponse {
        models: Vec<GoogleModel>,
    }

    let models_response: GoogleModelsResponse = response.json().await?;

    let models: Vec<DiscoveredModel> = models_response
        .models
        .into_iter()
        .filter(|m| {
            // Filter to generative models (Gemini)
            m.name.contains("gemini")
        })
        .map(|m| {
            // Extract model ID from full name (e.g., "models/gemini-pro" -> "gemini-pro")
            let id = m.name.strip_prefix("models/").unwrap_or(&m.name).to_string();
            DiscoveredModel {
                id: id.clone(),
                name: m.display_name.unwrap_or(id),
                description: m.description,
                context_window: m.input_token_limit,
                max_output_tokens: m.output_token_limit,
                owned_by: Some("google".to_string()),
                created: None,
            }
        })
        .collect();

    Ok(models)
}

File: packages/core/src/models/discovery.ts
/**
 * Model discovery module
 *
 * Discovers models from provider APIs.
 */
import type { Model } from "../types";
import type { ProviderType } from "../types";
import { providerManager } from "../providers";
import { proxyFetch } from "../http";

/**
 * Discovered model from an API
 */
export interface DiscoveredModel {
  /** Model ID */
  id: string;
  /** Display name */
  name?: string;
  /** Creation timestamp (if available) */
  created_at?: number;
  /** Owner/organization */
  owned_by?: string;
  /** Model capabilities */
  capabilities?: string[];
  /** Additional metadata */
  metadata?: Record<string, unknown>;
}

/**
 * Discovery result
 */
export interface DiscoveryResult {
  /** Provider ID */
  providerId: string;
  /** Provider type */
  providerType: ProviderType;
  /** Discovered models */
  models: DiscoveredModel[];
  /** Error if discovery failed */
  error?: string;
}

/**
 * OpenAI models response type
 */
interface OpenAIModelsResponse {
  data: Array<{
    id: string;
    created?: number;
    owned_by?: string;
  }>;
}

/**
 * Discover models from OpenAI API
 */
async function discoverOpenAI(
  apiKey: string,
  baseUrl = "https://api.openai.com/v1"
): Promise<DiscoveredModel[]> {
  const response = await proxyFetch(`${baseUrl}/models`, {
    headers: {
      Authorization: `Bearer ${apiKey}`,
    },
  });

  if (!response.ok) {
    throw new Error(`OpenAI API error: ${response.status} ${response.statusText}`);
  }

  const data = (await response.json()) as OpenAIModelsResponse;
  const models: DiscoveredModel[] = [];

  for (const model of data.data || []) {
    models.push({
      id: model.id,
      name: model.id,
      created_at: model.created,
      owned_by: model.owned_by,
    });
  }

  return models;
}

/**
 * Discover models from Anthropic API
 *
 * Note: Anthropic doesn't have a models list endpoint.
 */
async function discoverAnthropic(_apiKey: string, _baseUrl?: string): Promise<DiscoveredModel[]> {
  return [];
}

/**
 * Ollama models response type
 */
interface OllamaModelsResponse {
  models: Array<{
    name: string;
    size?: number;
    digest?: string;
    modified_at?: string;
  }>;
}

/**
 * Discover models from Ollama
 */
async function discoverOllama(baseUrl = "http://localhost:11434"): Promise<DiscoveredModel[]> {
  const response = await proxyFetch(`${baseUrl}/api/tags`);

  if (!response.ok) {
    throw new Error(`Ollama API error: ${response.status} ${response.statusText}`);
  }

  const data = (await response.json()) as OllamaModelsResponse;
  const models: DiscoveredModel[] = [];

  for (const model of data.models || []) {
    models.push({
      id: model.name,
      name: model.name,
      metadata: {
        size: model.size,
        digest: model.digest,
        modifiedAt: model.modified_at,
      },
    });
  }

  return models;
}

/**
 * OpenRouter models response type
 */
interface OpenRouterModelsResponse {
  data: Array<{
    id: string;
    name?: string;
    context_length?: number;
    pricing?: unknown;
    architecture?: unknown;
  }>;
}

/**
 * Discover models from OpenRouter
 */
async function discoverOpenRouter(apiKey: string): Promise<DiscoveredModel[]> {
  const response = await proxyFetch("https://openrouter.ai/api/v1/models", {
    headers: {
      Authorization: `Bearer ${apiKey}`,
    },
  });

  if (!response.ok) {
    throw new Error(`OpenRouter API error: ${response.status} ${response.statusText}`);
  }

  const data = (await response.json()) as OpenRouterModelsResponse;
  const models: DiscoveredModel[] = [];

  for (const model of data.data || []) {
    models.push({
      id: model.id,
      name: model.name || model.id,
      metadata: {
        contextLength: model.context_length,
        pricing: model.pricing,
        architecture: model.architecture,
      },
    });
  }

  return models;
}

/**
 * Google AI models response type
 */
interface GoogleAIModelsResponse {
  models: Array<{
    name?: string;
    displayName?: string;
    supportedGenerationMethods?: string[];
    description?: string;
    inputTokenLimit?: number;
    outputTokenLimit?: number;
    temperature?: number;
    topP?: number;
    topK?: number;
  }>;
}

/**
 * Discover models from Google AI (Gemini)
 */
async function discoverGoogle(apiKey: string): Promise<DiscoveredModel[]> {
  const response = await proxyFetch(
    `https://generativelanguage.googleapis.com/v1beta/models?key=${apiKey}`
  );

  if (!response.ok) {
    throw new Error(`Google AI API error: ${response.status} ${response.statusText}`);
  }

  const data = (await response.json()) as GoogleAIModelsResponse;
  const models: DiscoveredModel[] = [];

  for (const model of data.models || []) {
    models.push({
      id: model.name?.replace("models/", "") || model.name || "",
      name: model.displayName,
      capabilities: model.supportedGenerationMethods,
      metadata: {
        description: model.description,
        inputTokenLimit: model.inputTokenLimit,
        outputTokenLimit: model.outputTokenLimit,
        temperature: model.temperature,
        topP: model.topP,
        topK: model.topK,
      },
    });
  }

  return models;
}

/**
 * Discover models from Azure OpenAI
 *
 * Azure doesn't have a public models list endpoint.
 * Deployments are user-specific — returns empty.
 */
async function discoverAzure(): Promise<DiscoveredModel[]> {
  return [];
}

/**
 * Discover models from a provider
 *
 * @param providerId - The provider ID to discover models from
 * @returns Discovery result with models or error
 */
export async function discoverModels(providerId: string): Promise<DiscoveryResult> {
  const provider = await providerManager.getProvider(providerId);
  if (!provider) {
    return {
      providerId,
      providerType: "custom",
      models: [],
      error: `Provider not found: ${providerId}`,
    };
  }

  const providerType = provider.type as ProviderType;
  const apiKey = provider.apiKey || "";
  const baseUrl = provider.base_url;

  try {
    let models: DiscoveredModel[] = [];

    switch (providerType) {
      case "openai":
        models = await discoverOpenAI(apiKey, baseUrl);
        break;
      case "anthropic":
        models = await discoverAnthropic(apiKey, baseUrl);
        break;
      case "ollama":
        models = await discoverOllama(baseUrl || "http://localhost:11434");
        break;
      case "openrouter":
        models = await discoverOpenRouter(apiKey);
        break;
      case "google":
        models = await discoverGoogle(apiKey);
        break;
      case "azure":
        models = await discoverAzure();
        break;
      case "custom":
        if (baseUrl) {
          models = await discoverOpenAI(apiKey, baseUrl);
        }
        break;
      default:
        return {
          providerId,
          providerType,
          models: [],
          error: `Unsupported provider type: ${providerType}`,
        };
    }

    return {
      providerId,
      providerType,
      models,
    };
  } catch (e) {
    return {
      providerId,
      providerType,
      models: [],
      error: e instanceof Error ? e.message : String(e),
    };
  }
}

/**
 * Discover models from all configured providers
 *
 * @returns Array of discovery results for each provider
 */
export async function discoverAllModels(): Promise<DiscoveryResult[]> {
  const providers = await providerManager.listProviders();
  const results: DiscoveryResult[] = [];

  for (const provider of providers) {
    const result = await discoverModels(provider.id);
    results.push(result);
  }

  return results;
}

/**
 * Enrich discovered models with stored model information
 */
export function enrichModel(discovered: DiscoveredModel): Model | DiscoveredModel {
  return discovered;
}
